cmd/collectors/rest/plugins/svm/svm.go

```diff
@@ -5,43 +5,82 @@
 package svm
 
 import (
+	"github.com/tidwall/gjson"
+	"goharvest2/cmd/collectors"
 	"goharvest2/cmd/poller/plugin"
 	"goharvest2/pkg/matrix"
 	"strings"
 )
 
 type SVM struct {
 	*plugin.AbstractPlugin
+	nsswitchInfo map[string]nsswitch
+}
+
+type nsswitch struct {
+	nsdb     []string
+	nssource []string
 }
 
 func New(p *plugin.AbstractPlugin) plugin.Plugin {
 	return &SVM{AbstractPlugin: p}
 }
 
 func (my *SVM) Run(data *matrix.Matrix) ([]*matrix.Matrix, error) {
-	my.setNameservice(data)
+	var (
+		err error
+	)
+
+	// invoke nameservice-nsswitch-get-iter zapi and get nsswitch info
+	if my.nsswitchInfo, err = my.GetNSSwitchInfo(data); err != nil {
+		my.Logger.Warn().Stack().Err(err).Msg("Failed to collect nsswitch info")
+		//return nil, nil
+	}
+
+	// update svm instance based on the above zapi response
+	for _, svmInstance := range data.GetInstances() {
+		svmName := svmInstance.GetLabel("svm")
+
+		// Update nameservice_switch and nis_domain label in svm
+		if nsswitchInfo, ok := my.nsswitchInfo[svmName]; ok {
+			ns_db := strings.Join(nsswitchInfo.nsdb, ",")
+			ns_source := strings.Join(nsswitchInfo.nssource, ",")
+			nis_domain := svmInstance.GetLabel("nis_domain")
+			svmInstance.SetLabel("ns_source", ns_source)
+			svmInstance.SetLabel("ns_db", ns_db)
+			collectors.SetNameservice(ns_db, ns_source, nis_domain, svmInstance)
+		}
+	}
 	return nil, nil
 }
 
-func (my *SVM) setNameservice(data *matrix.Matrix) {
-	for _, instance := range data.GetInstances() {
-		requiredNSDb := false
-		requiredNSSource := false
-		ns := instance.GetLabel("nameservice_switch")
-		nisDomain := instance.GetLabel("nis_domain")
-
-		if strings.Contains(ns, "passwd") || strings.Contains(ns, "group") || strings.Contains(ns, "netgroup") {
-			requiredNSDb = true
-			if strings.Contains(ns, "nis") {
-				requiredNSSource = true
-			}
-		}
+func (my *SVM) GetNSSwitchInfo(data *matrix.Matrix) (map[string]nsswitch, error) {
 
-		if nisDomain != "" && requiredNSDb && requiredNSSource {
-			instance.SetLabel("nis_authentication_enabled", "true")
-		} else {
-			instance.SetLabel("nis_authentication_enabled", "false")
-		}
+	var (
+		vserverNsswitchMap map[string]nsswitch
+		ns                 nsswitch
+		ok                 bool
+	)
+
+	vserverNsswitchMap = make(map[string]nsswitch)
+
+	for _, svmInstance := range data.GetInstances() {
+		svmName := svmInstance.GetLabel("svm")
+		nsswitchConfig := svmInstance.GetLabel("nameservice_switch")
 
+		config := gjson.Result{Type: gjson.JSON, Raw: nsswitchConfig}
+
+		for nsdb, nssource := range config.Map() {
+			nssourcelist := strings.ReplaceAll(strings.TrimSuffix(strings.TrimPrefix(nssource.String(), "["), "]"), "\"", "")
```

Comment:
would it be a better fit to create a `strings.NewReplacer()` outside the loop and use it here?

Reply:
done